File: lg/src/questions/q_0105_construct_binary_tree_from_preorder_and_inorder_traversal.go
package questions

import (
	"github.com/hi-rustin/lg/src/utils"
)

func buildTree(preorder []int, inorder []int) *utils.TreeNode {
	if len(preorder) == 0 || len(inorder) == 0 {
		return nil
	}
	root := &utils.TreeNode{
		Val: preorder[0],
	}
	mid := 0
	for i, v := range inorder {
		if v == preorder[0] {
			mid = i
		}
	}

	root.Left = buildTree(preorder[1:mid+1], inorder[:mid])
	root.Right = buildTree(preorder[mid+1:], inorder[mid+1:])

	return root
}

File: lr/src/questions/q_0105_construct_binary_tree_from_preorder_and_inorder_traversal.rs
use std::{cell::RefCell, rc::Rc};

use crate::utils::TreeNode;

pub fn build_tree(preorder: Vec<i32>, inorder: Vec<i32>) -> Option<Rc<RefCell<TreeNode>>> {
    fn helper(preorder: &[i32], inorder: &[i32]) -> Option<Rc<RefCell<TreeNode>>> {
        if preorder.is_empty() || inorder.is_empty() {
            return None;
        }

        let mut root = Some(Rc::new(RefCell::new(TreeNode::new(preorder[0]))));
        let mid = inorder.iter().position(|&n| n == preorder[0]).unwrap();

        root.as_mut().unwrap().borrow_mut().left = helper(&preorder[1..=mid], &inorder[..mid]);
        root.as_mut().unwrap().borrow_mut().right =
            helper(&preorder[mid + 1..], &inorder[mid + 1..]);
        root
    }
    helper(&preorder, &inorder)
}

#[test]
fn test_build_tree() {
    use crate::tree;
    assert_eq!(
        build_tree(vec![3, 9, 20, 15, 7], vec![9, 3, 15, 20, 7]),
        tree!(3, 9, 20, None, None, 15, 7)
    );
    assert_eq!(build_tree(vec![1], vec![1]), tree!(1));
}
